src/uct/ib/rdmacm/rdmacm_ep.c

```diff
@@ -8,10 +8,57 @@
 ucs_status_t uct_rdmacm_ep_resolve_addr(uct_rdmacm_ep_t *ep)
 {
     uct_rdmacm_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_rdmacm_iface_t);
+    ucs_status_t status;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+
+    status = uct_rdmacm_resolve_addr(ep->rdmacm_cm_id->cm_id,
+                                    (struct sockaddr *)&(ep->remote_addr),
+                                    UCS_MSEC_PER_SEC * iface->config.addr_resolve_timeout,
+                                    UCS_LOG_LEVEL_ERROR);
 
-    return uct_rdmacm_resolve_addr(iface->cm_id, (struct sockaddr *)&(ep->remote_addr),
-                                   UCS_MSEC_PER_SEC * iface->config.addr_resolve_timeout,
-                                   UCS_LOG_LEVEL_ERROR);
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+    return status;
+}
+
+ucs_status_t uct_rdmacm_ep_set_cm_id(uct_rdmacm_iface_t *iface, uct_rdmacm_ep_t *ep)
+{
+    ucs_status_t status;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+
+    /* create a cm_id for the client side */
+    if ((iface->num_cm_id_in_quota <= iface->config.cm_id_quota_size) &&
+        (iface->num_cm_id_in_quota > 0)) {
+        /* Create an id for this interface. Events associated with this id will be
+         * reported on the event_channel that was created on iface init. */
+        ep->rdmacm_cm_id = ucs_malloc(sizeof(uct_rdmacm_cm_id_t), "client rdmacm_cm_id");
+        if (ep->rdmacm_cm_id == NULL) {
+            status = UCS_ERR_NO_MEMORY;
+            goto out;
+        }
+        if (rdma_create_id(iface->event_ch, &ep->rdmacm_cm_id->cm_id,
+                           ep->rdmacm_cm_id, RDMA_PS_UDP)) {
+            ucs_error("rdma_create_id() failed: %m");
+            ucs_free(ep->rdmacm_cm_id);
+            status = UCS_ERR_IO_ERROR;
+            goto out;
+        }
+        ep->rdmacm_cm_id->ep = ep;
+        ucs_list_add_tail(&iface->used_cm_ids_list, &ep->rdmacm_cm_id->list);
+        ep->is_on_cm_ids_list = 1;
+        iface->num_cm_id_in_quota--;
+        ucs_debug("ep %p, new cm_id %p. cm_id_in_quota %d", ep,
+                   ep->rdmacm_cm_id->cm_id, iface->num_cm_id_in_quota);
+        status = UCS_OK;
+    } else {
+        ep->is_on_cm_ids_list = 0;
```

Comment:
is to possible to set ep->rdmacm_cm_id to NULL to indicate it's not present, instead of adding another field 'ep->is_on_cm_ids_list'

Reply:
i think it's possible but then in the uct_rdmacm_iface_release_cm_id function, the code will appear:

```
    if (rdmacm_cm_id->ep != NULL) {
        rdmacm_cm_id->ep->rdmacm_cm_id = NULL;
    }
    rdma_destroy_id(rdmacm_cm_id->cm_id);
    ucs_free(rdmacm_cm_id);
```

seems a bit off isn't it?

Reply:
i think it's ok